src/HPolyhedron.jl

```diff
@@ -411,47 +419,88 @@ Concrete linear map of a polyhedron in constraint representation.
+linear map to each vertex.
+
+To switch off the check for invertibility, set the option
+`check_invertibility=false`. If `M` is not invertible and the polyhedron
+is unbounded, this function returns an exception.
 
-The option `use_inv` is a workaround to allow using the invertibility condition
-when `M` is a sparse matrix, since the `inv` function is not available for
-sparse matrices. In that case, either assure that `use_inv=false`, or use
-`linear_map(Matrix(M), P)`.
+The option `use_inv` lets the user control - in case `M` is invertible - if
+the full matrix inverse is computed, or only the left division on the normal
+vectors. Note that this helps as a workaround when `M` is a sparse matrix, since
+the `inv` function is not available for sparse matrices. In this case, either
+use the option `use_inv=false` or convert the type of `M` as in `linear_map(Matrix(M), P)`.
 """
 function linear_map(M::AbstractMatrix{N},
-                    P::PT;
+                    P::AbstractPolyhedron{N};
+                    check_invertibility::Bool=true,
                     cond_tol::Number=DEFAULT_COND_TOL,
                     use_inv::Bool=!issparse(M)
-                   ) where {N<:Real, PT<:HPoly{N}}
-    if !isinvertible(M; cond_tol=cond_tol)
-        if P isa HPolyhedron
-            error("linear maps for polyhedra need to be invertible")
-        end
-        # use the implementation for general polytopes
-        return invoke(linear_map, Tuple{typeof(M), AbstractPolytope{N}}, M, P)
-    end
+                   ) where {N<:Real}
+    @assert dim(P) == size(M, 2) "a linear map of size $(size(M)) cannot be applied to a set of dimension $(dim(P))"
 
-    constraints = similar(constraints_list(P))
+    if !check_invertibility || !isinvertible(M; cond_tol=cond_tol)
+        return _linear_map_vrep(M, P)
+    end
 
     # matrix M is invertible => the normal vectors are vec(c.a' * inv(M))
+    return _linear_map_hrep(M, P, use_inv)
+end
+
+function _linear_map_vrep(M::AbstractMatrix{N}, P::AbstractPolyhedron{N}) where {N<:Real}
+    if !isbounded(P)
+        throw(ArgumentError("the linear map in vertex representation for an unbounded set is not defined"))
+    else
+        # since P is bounded, we pass an HPolytope and then convert it to vertex representation
+        P = tovrep(HPolytope(constraints_list(P), check_boundedness=false))
+    end
+    return _linear_map_vrep(M, P)
+end
+
+function _linear_map_hrep(M::AbstractMatrix{N}, P::AbstractPolyhedron{N},
+                          use_inv::Bool) where {N<:Real}
+    constraints = _linear_map_hrep_helper(M, P, use_inv)
+    return HPolyhedron(constraints)
+end
+
+function _linear_map_hrep(M::AbstractMatrix{N}, P::AbstractPolytope{N},
+                          use_inv::Bool) where {N<:Real}
+    constraints = _linear_map_hrep_helper(M, P, use_inv)
+    return dim(P) == 2 ? HPolygon(constraints) : HPolytope(constraints)
```

Comment:
I begin to not like this line. If the user works with `HPolytope`s, it would be surprising if the result becomes an `HPolygon`. Let us add another method for `AbstractHPolygon` :blush:

Reply:
Again thanks for the good remark. Let's keep this method as fallback (`_linear_map_hrep` to dispatch on the general `AbstractPolytope`) but as you propose add two new methods to guarantee that the types are stable, one for `AbstractHPolygon` and another one for `HPolytope`.

Reply:
Alright, i have added two new methods to guarantee that the types are stable. In particular:

- the linear map of an `HPolytope` passes first through the `AbstractPolyhedron` interface `linear_map`; then goes to `_linear_map_vrep` and returns a `VPolygon` or a `VPolytope` depending on the dimension, and if h-rep is used returns `HPolytope`